packages/core/src/browser/shell/tab-bar-toolbar.ts

```diff
@@ -0,0 +1,245 @@
+/********************************************************************************
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { inject, injectable, named } from 'inversify';
+import { Widget, BaseWidget } from '../widgets';
+import { LabelParser, LabelIcon } from '../label-parser';
+import { ContributionProvider } from '../../common/contribution-provider';
+import { FrontendApplicationContribution } from '../frontend-application';
+import { Disposable, DisposableCollection } from '../../common/disposable';
+import { Command, CommandRegistry, CommandService } from '../../common/command';
+
+/**
+ * Factory for instantiating tab-bar toolbars.
+ */
+export const TabBarToolbarFactory = Symbol('TabBarToolbarFactory');
+export interface TabBarToolbarFactory {
+    (commandService: CommandService, labelParser: LabelParser): TabBarToolbar;
+}
+
+/**
+ * Tab-bar toolbar widget representing the active [tab-bar toolbar items](TabBarToolbarItem).
+ */
+export class TabBarToolbar extends BaseWidget {
+
+    protected readonly items: Map<TabBarToolbarItem, HTMLElement> = new Map();
+    protected readonly toDisposeOnUpdate: DisposableCollection = new DisposableCollection();
+
+    constructor(protected readonly commandService: CommandService, protected readonly labelParser: LabelParser) {
+        super();
+        this.toDispose.push(Disposable.create(() => this.removeItems()));
+        this.addClass(TabBarToolbar.Styles.TAB_BAR_TOOLBAR);
+        this.addClass(TabBarToolbar.Styles.TAB_BAR_TOOLBAR_HIDDEN);
+    }
+
+    updateItems(...items: TabBarToolbarItem[]): void {
+        const copy = items.slice().sort(TabBarToolbarItem.PRIORITY_COMPARATOR).reverse();
+        if (this.areSame(copy, Array.from(this.items.keys()))) {
+            return;
+        }
+        this.toDisposeOnUpdate.dispose();
+        this.removeItems();
+        this.createItems(...copy);
+    }
+
+    protected removeItems(): void {
+        for (const element of this.items.values()) {
+            const { parentElement } = element;
+            if (parentElement) {
+                parentElement.removeChild(element);
+            }
+        }
+        this.items.clear();
+    }
+
+    protected createItems(...items: TabBarToolbarItem[]): void {
```

Comment:
It might be faster than manipulating the DOM directly. Since the check for necessary DOM updates is done by the VirtualDOM class, the `areSame` check could be removed.

But this is not critical for this PR, it could be a follow-up optimization.